### PR TITLE
Release Candidate for Arizona scraper #9 + changes in Oragon scraper

### PR DESCRIPTION
**Pull request by Eiji7**
**Note**: I don't know if it much all your requirements, so I called it "Release Candidate".
**Note2**: I also don't have created tests. Just created scraper to check if it much your requirements.

### CODE REVIEW

File: lib/nh_verify/producers/arizona.ex [hunk 1]
```diff
@@ -0,0 +1,298 @@
...
+      "#{@base_url}#{href}"
+      |> get!
+      |> Floki.find("font")
+      |> List.first
+    [content | get_next_search_page(content, false)]
+  end
+
+  defp get_next_search_page(content, main_page \\ true) do
+    link = find_link(content, main_page)
+    if link, do: do_get_next_search_page(link), else: []
```
Comment (jeffdeville) [inline lib/nh_verify/producers/arizona.ex]: Usually, what I try to do is have my code read top to bottom. So do_get_next_search_page would be BELOW this one. But you seem to have looked more at the code standard docs than I have.  Is this the way they do it?

File: lib/nh_verify/producers/arizona.ex [hunk 2]
```diff
@@ -0,0 +1,298 @@
...
+    |> Stream.map(&to_licensee(&1))
+  end
+
+  def get!(url) do
+    {:ok, resp} = retry with: exp_backoff |> randomize |> expiry(10_000_000) do
+      HTTPoison.get(url, [], ssl: [versions: [:"tlsv1.2"]])
+    end
+    result =
+      resp.body
+      |> Codepagex.to_string!(:"ISO8859/8859-1")
```
Comment (jeffdeville) [inline lib/nh_verify/producers/arizona.ex]: Is this really needed here?  By default, we assume UTF-8, and there is a HEAD tag that claims that is what they're returning to us.

File: lib/nh_verify/producers/arizona.ex [hunk 3]
```diff
@@ -0,0 +1,298 @@
...
+    |> Stream.map(&remove_old_issues(&1))
+    |> Stream.map(&collect_data(&1))
+    |> Stream.filter(&valid_license_num?(&1))
+    |> Stream.map(&collect_data_from_details(&1))
+    |> Stream.map(&to_licensee(&1))
+  end
+
+  def get!(url) do
+    {:ok, resp} = retry with: exp_backoff |> randomize |> expiry(10_000_000) do
+      HTTPoison.get(url, [], ssl: [versions: [:"tlsv1.2"]])
```
Comment (jeffdeville) [inline lib/nh_verify/producers/arizona.ex]: The ssl flag isn't needed for http resources.

File: lib/nh_verify/producers/arizona.ex [hunk 4]
```diff
@@ -0,0 +1,298 @@
...
+  end
+
+  defp get_main_search_page do
+    "#{@base_url}directorySearch.asp?licenseType=1"
+    |> get!
+    |> Floki.find("font")
+    |> List.first
+  end
+
+  defp find_link({_, _, children}, main_page) do
```
Comment (jeffdeville) [inline lib/nh_verify/producers/arizona.ex]: I think I was probably guilty of similarly unclear naming in oregon.ex, but could you rename this method to make it clear what type of link this is looking for?  Paging links?

File: lib/nh_verify/producers/arizona.ex [hunk 5]
```diff
@@ -0,0 +1,298 @@
...
+    links =
+      children
+      |> Enum.fetch!(1)
+      |> Floki.find("font > b > a")
+    links_count = Enum.count(links)
+    cond do
```
Comment (jeffdeville) [inline lib/nh_verify/producers/arizona.ex]: I think you could write this in a way to save a few lines:

```
cond Enum.count(links) do
  4 -> Enum.fetch!(links, 2)
  2 -> if main_page, do: List.first(links), else: nil
  _ -> nil
end
```

I don't see how True would ever be a return value here.

File: lib/nh_verify/producers/arizona.ex [hunk 6]
```diff
@@ -0,0 +1,298 @@
...
+        Enum.fetch!(links, 2)
+      links_count == 2 ->
+        if main_page, do: List.first(links), else: nil
+      true ->
+        nil
+    end
+  end
+
+  defp do_get_next_search_page(link) do
+    href =
```
Comment (jeffdeville) [inline lib/nh_verify/producers/arizona.ex]: I'm not requesting you do this unless you want, but knowing you like short files, I think you could have put this on 1 line and it would have been just as readable.  If you did this throughout the file for short |> groups, I think you could probably save 30 lines.

`href = link |> Floki.attribute("href") |> hd`

I altered the credo settings to permit line lengths of up to 120 characters, FYI.

File: lib/nh_verify/producers/arizona.ex (hunk 3 above)
Comment (Eiji7) [inline lib/nh_verify/producers/arizona.ex]: I just copied your method from first scraper 😄, but ok I will correct it.

File: lib/nh_verify/producers/arizona.ex (hunk 2 above)
Comment (Eiji7) [inline lib/nh_verify/producers/arizona.ex]: same as before: I just copied it from your first scraper 😄, but ok I will correct it.

File: lib/nh_verify/producers/arizona.ex (hunk 4 above)
Comment (Eiji7) [inline lib/nh_verify/producers/arizona.ex]: I'm looking for next page link, I will correct it.

File: lib/nh_verify/producers/arizona.ex (hunk 5 above)
Comment (Eiji7) [inline lib/nh_verify/producers/arizona.ex]: You mean **case** - **cond** statement works different and true is like **else** here, but I agree that I can change it to **case**.

File: lib/nh_verify/producers/arizona.ex (hunk 1 above)
Comment (Eiji7) [inline lib/nh_verify/producers/arizona.ex]: I don't remember any rule for that . I will search and I will post here more info with link.

**Comment by jeffdeville** [comment]
I just pushed a PR to update the database structure.  Sorry I didn't have too much time today. I'm falling behind with my day job a bit. Can you explain what you're struggling w/ regarding testing? I'll see if I can help.

File: lib/nh_verify/producers/arizona.ex [hunk 7]
```diff
@@ -0,0 +1,292 @@
...
+  @behaviour NHVerify.Producer
+
+  @base_url "http://directorynd.az.gov/agency/pages/"
+  @first_license_list_url "#{@base_url}directorySearch.asp?licenseType=1"
+  @main_content_selector "#node-161 > .node-content > .content"
+
+  def get_licensees() do
+    @first_license_list_url
+    |> Stream.unfold(&get_search_page_contents/1)
+    |> Stream.map(&remove_table_headers/1)
```
Comment (jeffdeville) [inline lib/nh_verify/producers/arizona.ex]: Lines 15-18 could be: converted to a higher-level method called: filter_page_garbage
Lines 19-23 could be converted to a higher level method called: extract_license

**Comment by Eiji7** [comment]
I added tests (also for edge cases). I think I done this scraper well.
